src/compiler/checker.ts

```diff
@@ -17883,9 +17883,11 @@ namespace ts {
                     const template = getTemplateTypeFromMappedType(target);
                     const modifiers = getMappedTypeModifiers(target);
                     if (!(modifiers & MappedTypeModifiers.ExcludeOptional)) {
-                        if (template.flags & TypeFlags.IndexedAccess && (<IndexedAccessType>template).objectType === source &&
-                            (<IndexedAccessType>template).indexType === getTypeParameterFromMappedType(target)) {
-                            return Ternary.True;
+                        if (template.flags & TypeFlags.IndexedAccess && (<IndexedAccessType>template).indexType === getTypeParameterFromMappedType(target)) {
+                            const objectType = (<IndexedAccessType>template).objectType;
+                            if (objectType === source) return Ternary.True;
+                            const objectTypeConstraint = getConstraintOfType(objectType);

```

Comment:
I think you don't want to look at the target's constraints at all - knowing that e.g. `T` is a subtype of `{x: string}` doesn't help you assign anything to a `T`, since `T` could have stricter requirements (like `x: "yes" | "no"`).

We don't actually care about any of its attributes/properties for the assignability relation. The only reason we even need the `T extends { foo: string }` constraint is so that it's possible to write `T["foo"]` in the signature, which is handled elsewhere. So I think the correct/complete version of this code won't look at the constraint type of the target _at all_.

---

Note, here's an example of a function that is currently rejected, _and should continue to be rejected_ after this change. It doesn't have a `extends` clause on `T`, but that's not the reason for it:

```ts
function shouldReject<T, K extends keyof T>(key: K, v: T[K]): {[k in keyof T]?: T[k]} {
	return { [key]: v }
    // Type '{ [x: string]: T[K]; }' is not assignable to type '{ [k in keyof T]?: T[k] | undefined; }'.(2322)
}
```

in this case, we don't know whether `K` is just one value, or a union of possible values. For example, if we called `shouldReject<{foo: string, bar: number, qux: boolean}, "foo" | "bar">`  then `T[K]` is `string | number`, which means that we could pass in `"(foo", 123)` and produce a value that's clearly not a `Partial<T>`.

I actually proposed #41363 a while ago to deal with this; the correct typing for the function would be `T[assign K]` instead of `T[K]` in the argument, but that's clearly beyond the scope of this PR, it's just vaguely relevant.

Reply:
I agree with the more general point that it is rarely useful to look at the constraints of types when determining whether one is an instance of the other. I do think, however, that this is one of the rare cases where it actually is useful.
I completely agree with all the other points you're making 🙂
I also added your example to the test, although I don' think it will be immediately impacted by the change.